Feature/api security

danewrye opened the pull request:
- Added loci_repeats parameter to api
- Added custom exception handling for malformatted loci_repeats
- Updated API Readme to fix previous inconsistencies and add information about new parameters
- Added API Rate Limiting based on remote address (see library used at https://flask-limiter.readthedocs.io/en/stable/)

danewrye committed:
Update api md file

Added information on new params. Corrected errors regarding file type and http methods

danewrye committed:
Add new api parameters

Added loci_repeats and advanced_stats api parameters

danewrye committed:
Implement API Rate limiting

Implemented rate limiting on hourly, per minute, and per second basis for API

danewrye committed:
Remove Advanced Stats

As of last meeting, advanced stats is no longer needed

danewrye committed:
Added locirepeats error handling

Added exception and handling for malformatted locirepeats

JHDeerin commented on api/API.md:
```diff
@@ -39,29 +39,34 @@ The root endpoint for all API requests for version 1.0 of the API is `/api/v1`
--   `file` (.csv file, FormData): The file containing the testing data to be processed.
+-   `file` (.xlsx file, FormData): The file containing the testing data to be processed.
+
+**Input Parameters (URL):**
```
Don't create a 2nd "Input Parameters (URL)" section; instead, add this as a bullet point under the 1st "Input Parameters (URL)"

danewrye replied:
Makes sense, change made

JHDeerin commented on api/API.md:
```diff
@@ -39,29 +39,34 @@ The root endpoint for all API requests for version 1.0 of the API is `/api/v1`
+
+**Input Parameters (URL):**
+
+-   (optional) `locirepeat` (list of integers): List of integers representing loci repeats. Each loci repeat must be individually entered for the request. Please ensure the correct amount of values are entered.
```
Locirepeats shouldn't be optional; it should be a required parameter (file can't be processed correctly without it)

JHDeerin replied:
Also, here's a description from `define_alleles` about what locirepeats actually are that might be useful to include: `a vector of length <number of loci> with the type of each locus (dinucleotide, trinucleotide, etc. repeats)`

danewrye replied:
Done, and thanks for the description! It is much better than the one I had in there

JHDeerin commented on api/API.md:
````diff
@@ -39,29 +39,34 @@ The root endpoint for all API requests for version 1.0 of the API is `/api/v1`
-curl -F "file=@example.csv" http://localhost:5000/api/v1/recrudescences?iterations=10
+curl -F "file=@example.xlsx" "http://localhost:5000/api/v1/recrudescences?iterations=10&locirepeat=2&locirepeat=2&locirepeat=3"
 ```
+If you are using curl and are entering multiple query parameters, ensure you use quotes around the request
````
This should already be known by people who are using curl and can easily be looked up, so I don't think we need to include it; if you still want to, I would reformat it slightly to be in italics/etc. (to make it clear it's a side-note)

danewrye replied:
True. I will delete it.

JHDeerin commented on api/API.md:
```diff
@@ -39,29 +39,34 @@ The root endpoint for all API requests for version 1.0 of the API is `/api/v1`
+Given an input `.xlsx` file with [drug testing data in the proper format (TODO)](), return estimates of which patients specified in the file are reinfections or recrudescences.
 
-*Method(s):* `PUSH`
+*Method(s):* `POST`
 
 *URL Endpoint:* `/api/v1/recrudescences`
 
-#### PUSH
+#### POST
```
Thank you! Good catch!

danewrye replied:
No problem!

JHDeerin commented on api/api.py:
```diff
@@ -44,13 +50,22 @@ def too_large(e):
         '''
         uploaded_file = request.files.get('file')
-        iterations = request.args.get('iterations', default=10000, type=int)
+        iterations = request.args.get('iterations', default=50, type=int)
```
Change the default back to 10000 (this should be set on the frontend, but might as well do it for future API use)

danewrye replied:
Done. Of all things to forget to do...

JHDeerin commented on api/api.py:
```diff
@@ -44,13 +50,22 @@ def too_large(e):
-        iterations = request.args.get('iterations', default=10000, type=int)
+        iterations = request.args.get('iterations', default=50, type=int)
+        loci_repeats = request.args.getlist('locirepeat', type=int)
+        if not loci_repeats:
+            loci_repeats = [2,2,3,3,3,3,3]
```
Remove this; this `loci_repeats` value is specific to the Angola example Mat gave us, so I would make loci_repeats required and throw an error if it isn't included instead

danewrye replied:
Done. Good improvement over the specific default

JHDeerin commented on api/algorithm_site_instance.py:
```diff
@@ -143,6 +145,9 @@ def __init__(
         self.locinames = pd.unique(genotypedata_RR.columns[1:].str.split("_").str[0])
 
+        if len(self.locinames) > len(locirepeats):
+            raise ValueError("Loci repeats has an insufficient number of entries")
```
Could you make a custom exception for improper loci repeats in this file, and then use that? I know that's annoying, but ValueErrors can be thrown by other things that might lead to confusing error messages about loci repeats that have nothing to do with the loci

danewrye replied:
Change made, and no problem! Completely see the point about clarity, and the effort to implement was minimal. With this change, I think it should be ready for re-review

JHDeerin commented on api/api.py:
```diff
@@ -1,11 +1,14 @@
 
 from flask import Flask, request
 from flask_restful import Resource, Api
+from flask_limiter import Limiter
```
Is `flask_limiter` a new library that has to be installed via `pip`? If so, can you add that library to `API.md` at the top, under the install instructions?

danewrye replied:
Good catch, completely forgot to add

JHDeerin commented on api/api.py:
```diff
@@ -44,13 +50,22 @@ def too_large(e):
 
 class RecrudescenceTest(Resource):
+
+    decorators = [
```
Where are these decorators actually applied? Quickly checking, this list syntax doesn't seem to be one of the standard ways of applying limits to a flask function (https://flask-limiter.readthedocs.io/en/stable/)

Have you tested to make sure the limiter is actually being used right now?

danewrye replied:
Yes, it has been tested and works properly. If you also want to test it, just turn the max request limit down very low. There were problems with the standard ways outlined in the docs- if you want I can try to find the stackoverflow post i referenced in implementing that line

JHDeerin reviewed (changes requested):
Overall this looks good! There's a few tweaks to the documentation I think are needed and making locirepeats a required parameter seems necessary, and I want to double-check these rate limits actually are working, but the core seems solid.

danewrye committed:
Requested changes for pull request

Improved readme, added custom locirepeat error, fixed default iteration value, changed error handling for locirepeat

JHDeerin commented on api/api.py:
```diff
@@ -109,7 +110,7 @@ def _get_test_results_json(
             results = test_run.run_algorithm(iterations, burnin, record_interval)
-        except ValueError:
+        except LociRepeatError:
             return error_response('Loci repeats has an insufficient number of entries', 400)
```
Testing this, it doesn't appear to actually be caught/printed in the response; I just get a generic "A problem occurred while the server was processing this data." Also, this method is currently expected to just return the JSON by `post`, not the HTML status code as well (despite what the docstring comment says).

Could you hunt down what's causing that (possibly the error doesn't get caught and a different error is thrown in AlgorithmInstance/etc?), and update the `post` method to use the HTML code from this as well?

danewrye replied:
I fixed the error, but I'm not 100% sure what you mean with the HTML status codes. I removed the only line in _get_test_results_json that returned an html status code (un-necessary, could be handled in post). Is that what you were asking? I don't think there are any HTML status codes from this method that can be returned. Or did you just want me to edit the docstring to reflect this?

JHDeerin reviewed (changes requested):
Everything looks good except for the failing error message; fix that bug, and that's it!

danewrye committed:
Fixed error handling bug

Locirepeats error handling now displays on user side

JHDeerin reviewed (approved):
Great! Let's merge this in!
